iothub/service/src/DirectMethod/DirectMethodsClient.cs

```diff
@@ -0,0 +1,147 @@
+        /// <param name="cloudToDeviceMethod">Parameters to execute a direct method on the device.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>The <see cref="CloudToDeviceMethodResult"/>.</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="deviceId"/> is null.</exception>
+        /// <exception cref="ArgumentNullException">When <paramref name="cloudToDeviceMethod"/> is null.</exception>
+        /// <exception cref="ArgumentException">When <paramref name="deviceId"/> is empty or whitespace.</exception>
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-direct-methods"/>
+        public virtual Task<CloudToDeviceMethodResult> InvokeAsync(string deviceId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken = default)
+        {
+            Argument.RequireNotNullOrEmpty(deviceId, nameof(deviceId));
+            Argument.RequireNotNull(cloudToDeviceMethod, nameof(cloudToDeviceMethod));
+
+            return InvokeAsync(GetDeviceMethodUri(deviceId), cloudToDeviceMethod, cancellationToken);
+        }
+
+        /// <summary>
+        /// Interactively invokes a method on a module.
+        /// </summary>
+        /// <param name="deviceId">The device identifier for the target device.</param>
+        /// <param name="moduleId">The module identifier for the target module.</param>
+        /// <param name="cloudToDeviceMethod">Parameters to execute a direct method on the module.</param>
+        /// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+        /// <returns>The <see cref="CloudToDeviceMethodResult"/>.</returns>
+        /// <exception cref="ArgumentNullException">When <paramref name="deviceId"/> or <paramref name="moduleId"/> are null.</exception>
+        /// <exception cref="ArgumentNullException">When <paramref name="cloudToDeviceMethod"/> is null.</exception>
+        /// <exception cref="ArgumentException">When <paramref name="deviceId"/> or <paramref name="moduleId"/> are empty or whitespace.</exception>
+        /// <seealso href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-direct-methods"/>
+        public virtual Task<CloudToDeviceMethodResult> InvokeAsync(string deviceId, string moduleId, CloudToDeviceMethod cloudToDeviceMethod, CancellationToken cancellationToken = default)
+        {
+            Argument.RequireNotNullOrEmpty(deviceId, nameof(deviceId));
+            Argument.RequireNotNullOrEmpty(moduleId, nameof(moduleId));
+            Argument.RequireNotNull(cloudToDeviceMethod, nameof(cloudToDeviceMethod));
+
+            return InvokeAsync(GetModuleMethodUri(deviceId, moduleId), cloudToDeviceMethod, cancellationToken);
+        }
+
+        private async Task<CloudToDeviceMethodResult> InvokeAsync(Uri uri,CloudToDeviceMethod cloudToDeviceMethod,CancellationToken cancellationToken)
```

Comment:
This private function can be deleted and its contents can be moved into the most overloaded public function. This helps to limit the call stack height when someone calls the least overloaded public function.